exchanges/request/limit.go

```diff
@@ -133,44 +124,59 @@ func (r *Requester) GetRateLimiterDefinitions() RateLimitDefinitions {
 	return r.limiter
 }
 
-// RateLimit is a function that will rate limit a request based on the rate
-// limiter provided. It will return an error if the context is cancelled or
-// deadline exceeded.
-func RateLimit(ctx context.Context, rateLimiter *RateLimiterWithWeight) error {
-	if rateLimiter == nil {
-		return errSpecificRateLimiterIsNil
+// RateLimit throttles a request based on weight, delaying the request
+// Errors if no delay is permitted via the context and a delay is required
+func (r *RateLimiterWithWeight) RateLimit(ctx context.Context) error {
+	if err := common.NilGuard(r); err != nil {
+		return err
+	}
+
+	if r.weight == 0 {
+		return errInvalidWeight
 	}
 
-	if rateLimiter.Weight <= 0 {
-		return errInvalidWeightCount
+	r.m.Lock()
+	tn := time.Now()
+	reserved := make([]*rate.Reservation, 0, r.weight)
+	for range r.weight {
+		// This avoids needing burst capacity in the limiter, which would otherwise allow the rate limit to be exceeded over short periods
+		reserved = append(reserved, r.limiter.ReserveN(tn, 1))
 	}
+	finalDelay := reserved[len(reserved)-1].DelayFrom(tn)
 
-	var finalDelay time.Duration
-	reservations := make(Reservations, rateLimiter.Weight)
-	for i := Weight(0); i < rateLimiter.Weight; i++ {
-		// Consume 1 weight at a time as this avoids needing burst capacity in the limiter,
-		// which would otherwise allow the rate limit to be exceeded over short periods
-		reservations[i] = rateLimiter.Reserve()
-		finalDelay = reservations[i].Delay()
+	if finalDelay == 0 {
+		r.m.Unlock()
+		return nil
 	}
 
-	if dl, ok := ctx.Deadline(); ok && dl.Before(time.Now().Add(finalDelay)) {
-		reservations.CancelAll()
-		return fmt.Errorf("rate limit delay of %s will exceed deadline: %w",
-			finalDelay,
-			context.DeadlineExceeded)
+	if hasDelayNotAllowed(ctx) {
+		cancelAll(reserved, tn)
```

Comment:
I couldn't see where this tn is updated ... it still represents time.Now() and it is being set as a Reservation cancellation timestamp